### PR TITLE
Refactoring the loading of XDP modules in XRT

### PR DESCRIPTION
**Pull request by jvillarre**
Modifying the loading of the hal_api_interface and hal_profiling XDP plugin modules to use the same common XDPLoader class as the XDP plugin modules loaded by the OpenCL level.  This required moving the XDPLoader class functionality into the xrt_coreutil library, which necessitated adding the "dl" library during linking on Linux.

### CODE REVIEW

**Comment by jvillarre** [comment]
I changed the names of the files, class, and include guards to better reflect that this is a generic module loader instead of XDP specific.  Also, added comments to describe intended usage and what the constructor parameters should be used for.

**Comment by jvillarre** [comment]
retest this please

File: src/runtime_src/core/common/module_loader.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+   * 
+   * This class is responsible for loading a plugin module from the
+   * appropriate directory under the XILINX_XRT installation.  The loading
+   * happens at object construction time, so the XRT side implementation should
+   * contain a function that instantiates a single static instance of this 
+   * class to handle the loading of a module once in a thread safe manner.
+   *
+   */
+
+  class module_loader
```
Comment (stsoe) [inline src/runtime_src/core/common/module_loader.h:42]: Why is this a class?

File: src/runtime_src/core/common/module_loader.h [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  {
+  private:
+    // Useful functions
+    const char* emptyOrValue(const char* cstr) ;
+    boost::filesystem::path& dllExt() ;
+    bool isDLL(const boost::filesystem::path& path) ;
+    boost::filesystem::path modulePath(const boost::filesystem::path& root,
+				       const std::string& libname);
+    boost::filesystem::path moduleDir(const boost::filesystem::path& root) ;
```
Comment (stsoe) [inline src/runtime_src/core/common/module_loader.h:51]: These looks like they should be static functions in unnamed namespace in module_loader.cpp

File: src/runtime_src/core/common/module_loader.h [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+ * under the License.
+ */
+
+#ifndef xrtcore_util_module_loader_h_
+#define xrtcore_util_module_loader_h_
+
+// This file contains a loader utility class for plugin modules
+//  that are loaded from either OpenCL or XRT level applications.
+
+#include <boost/filesystem/path.hpp>
```
Comment (stsoe) [inline src/runtime_src/core/common/module_loader.h:23]: This include should be in the .cpp file.

File: src/runtime_src/core/common/module_loader.h (hunk 2 above)
Comment (jvillarre) [reply src/runtime_src/core/common/module_loader.h]: The utility functions were changed into static functions in an unnamed namespace in the .cpp file.

File: src/runtime_src/core/common/module_loader.h (hunk 3 above)
Comment (jvillarre) [reply src/runtime_src/core/common/module_loader.h]: This include was moved into the cpp file along with the utility functions

File: src/runtime_src/core/common/module_loader.h (hunk 1 above)
Comment (jvillarre) [reply src/runtime_src/core/common/module_loader.h:41]: The loading of the plugin module should occur only once and is triggered by the first call into XRT source code from the user's side.  All of the entry points specific to a plugin will call a single function to load that plugin module.  In order to handle multithreaded access from the user and guarantee that the loading only occurs once, the loading functions consist of a single static declaration of an instance of this class.  This instance is only constructed once and is thread safe per C++-11, and simplifies the code by not having to create global locks for each plugin and check flags on every callback to see if loading should occur.

Comment (stsoe) [reply src/runtime_src/core/common/module_loader.h:41]: ok. makes sense

File: src/runtime_src/core/common/module_loader.h (hunk 2 above)
Comment (stsoe) [reply src/runtime_src/core/common/module_loader.h]: Thanks

**Comment by jvillarre** [comment]
retest this please

**Pull request merged**